lint: update commit message to include issue/epic refs

nickvigilante opened the pull request:
The standard commit message now includes some help text on including issue and epic references.

Fixes #90303 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90776)
<!-- Reviewable:end -->

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
+# Add an issue or epic reference.
+if ! grep -q -E '^(?i:close[sd]?|fix(?:e[sd])?|resolve[sd]?)' "$1"; then
+	sed_script+="/$cchar Please enter the commit message for your changes./i\\
+$cchar Please enter a valid issue or epic reference:\\
```
Most of the time our engineers will just know that an issue / epic ref is required and they just need a reminder. So this line doesn't proved a lot of extra value. Let's remove it and instead add text like the following after [this line](https://github.com/cockroachdb/cockroach/blob/421aff1fabcfe16d93ebf2e9169d19fa04c0c997/githooks/prepare-commit-msg#L123):

```
#cchar An issue or epic reference must be in the PR body or each commit message\\
#cchar if the PR or commit is part of an issue or epic. Use `Epic: none` otherwise.\\
#cchar See also: https://wiki.crdb.io/wiki/spaces/CRDB/pages/2009039063/\\
#cchar\\
```

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
+	sed_script+="/$cchar Please enter the commit message for your changes./i\\
+$cchar Please enter a valid issue or epic reference:\\
+${cchar}Fixes: #77376\\
+${cchar}Part of https://cockroachlabs.atlassian.net/browse/CRDB-13558\\
```
Suggesting these changes because issue references normally shouldn't be to CRDB issues because the cockroachdb/cockroach issues are the source of truth and it's easier to digest these suggestions if their formatting is the same.

```suggestion
${cchar}Part of: https://cockroachlabs.atlassian.net/browse/DOC-1355\\
```

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
+${cchar}Part of https://cockroachlabs.atlassian.net/browse/CRDB-13558\\
+${cchar}Informs: https://github.com/cockroachdb/cockroach/issues/33316\\
+${cchar}Epic: CRDB-8035\\
+${cchar}Epic: none\\
```
Let's put the `Epic: none` example first in the list. It's what people will copy/paste most often.

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
+${cchar}Informs: https://github.com/cockroachdb/cockroach/issues/33316\\
+${cchar}Epic: CRDB-8035\\
+${cchar}Epic: none\\
+${cchar}      ^-- the least preferred option\\
```
To tell them specifically when to use this option.

```suggestion
${cchar}.     ^-- not related to an issue or epic\\
```

nickvigilante replied:
@jlinder Is the period necessary here?

jlinder replied:
No.

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
+"
+fi
+
+# with '#' will be ignored
```
Is this a vestigial working note that can be removed?

nickvigilante replied:
Removed

jlinder reviewed (dismissed):
Cool. A couple comments.

- Please change the issue ref to reference the GitHub issue, not the Jira issue. The GitHub issues are the source of truth for CRDB Task/Bug issues.

nickvigilante commented on githooks/prepare-commit-msg:
```diff
@@ -96,6 +96,23 @@ ${cchar}Release note (bug fix): \\
 fi
 
+# Add an issue or epic reference.
+if ! grep -q -E '^(?i:close[sd]?|fix(?:e[sd])?|resolve[sd]?)' "$1"; then
```
Done. Grep doesn't like the `(?im)` at the beginning of our RegExes, so I omitted that, but I now check for the existence of one of the refs as closely to the original go script as I could get.

rail reviewed (approved):
I won't pretend that my brain parsed all of the regexes, but they look good to me.

nickvigilante commented:
bors r+

nickvigilante commented:
bors r+

nickvigilante commented:
bors r+

jlinder commented on githooks/prepare-commit-msg:
```diff
@@ -115,6 +142,9 @@ $cchar     <what was there before: Previously, ...>\\
 $cchar     <what you did about it: To address this, this patch ...>\\
 $cchar\\
+$cchar     Fixes <GH/Jira issue ID/URL to GH/Jira issue>\\
+$cchar     ---\\
```
Can we remove this line? It exists after the `Release note` to indicate the end of the sample commit message body.

nickvigilante replied:
Done.

Commit:
lint: update commit message to include issue/epic refs

The standard commit message now includes some help text on including
issue and epic references.

Fixes #90303

Release note: None

nickvigilante commented:
bors r+

rail commented:
bors retry

nickvigilante commented:
bors retry
